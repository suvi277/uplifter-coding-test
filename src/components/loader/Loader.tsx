import React from 'react';
import { Spinner } from 'react-bootstrap';
import './Loader.scss';

const Loader: React.FC<{}> = () => (
    <div className="loader">
        <Spinner animation="border" variant="warning" />
    </div>
);

export default Loader;
