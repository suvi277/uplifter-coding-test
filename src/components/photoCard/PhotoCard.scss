@import '../../_shared/styles/variable';
@import '../../_shared/styles/mixin';

.card {
    margin: 20px 15px;
    cursor: pointer;
    transition: transform 1s ease;
    width: 100%;

    .card-body {
        display: flex;
        flex-direction: column;
        justify-content: space-between;

        h5 {
            font-size: 14px;
        }

        p {
            margin: 0;
            display: inline;
            font-size: 12px;

            &:first-child,
            img {
                display: none;
            }
        }
    }

    &:hover {
        @include transform(scale(1.1));
    }

    .block-with-text {
        overflow: hidden;
        position: relative;
        line-height: 1.2em;
        max-height: 3.6em;
        margin-right: -1em;
        padding-right: 1em;
        width: 100%;
    }

    .block-with-text:before {
        content: '...';
        position: absolute;
        right: 0;
        bottom: 0;
    }

    .block-with-text:after {
        content: '';
        position: absolute;
        right: 0;
        width: 1em;
        height: 1em;
        margin-top: 0.2em;
        background: white;
    }

    @media screen and (min-width: 576px) {
        width: calc(100% / 2 - 30px);
        width: auto;
    }

    @media screen and (min-width: 768px) {
        width: calc(100% / 2 - 30px);
    }

    @media screen and (min-width: 1024px) {
        width: calc(100% / 4 - 30px);
    }
}
