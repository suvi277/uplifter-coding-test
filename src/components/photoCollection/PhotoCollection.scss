@import '../../_shared/styles/main';

.card-columns {
  @include media-breakpoint-only(lg) {
    column-count: 3;
  }
  @include media-breakpoint-only(xl) {
    column-count: 4;
  }
}