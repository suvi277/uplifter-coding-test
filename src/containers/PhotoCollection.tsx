import { connect } from 'react-redux';
import { State } from '../reducers';
import { getPhotos } from '../actions/photos/photos.action';
import PhotoCollectionComponent from '../components/photoCollection/PhotoCollection';

const mapStateToProps = (state: State) => ({ ...state.photoCollection });

const mapDispatchToProps = (dispatch: any) => ({ getPhotos: () => dispatch(getPhotos()) });

export const PhotoCollectionContainer = connect(mapStateToProps, mapDispatchToProps)(PhotoCollectionComponent);
