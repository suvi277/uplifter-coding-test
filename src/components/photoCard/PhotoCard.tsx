import React from 'react';
import { Photo } from '../../types/Photo';
import './PhotoCard.scss';
import Select from '../select/Select';

const PhotoCard: React.FC<Photo> = ({ title, ownername, description, url_q, owner, tags, id }) => {
    const alltags = tags ? tags.split(' ') : [];

    return (
        <div className="card shadow-lg">
            <img className="card-img-top" src={url_q} alt={title} />
            <div className="card-body">
                <h5 className="card-title">
                    <a
                        href={`https://www.flickr.com/photos/${owner}/${id}`}
                        rel="noopener noreferrer"
                        title={title}
                        target="_blank"
                    >
                        {title}
                    </a>{' '}
                    by{' '}
                    <a
                        href={`https://www.flickr.com/photos/${owner}`}
                        rel="noopener noreferrer"
                        title={ownername}
                        target="_blank"
                    >
                        {ownername}
                    </a>
                </h5>
                {description && (
                    <div
                        className="block-with-text"
                        dangerouslySetInnerHTML={{
                            __html: description._content
                        }}
                    />
                )}
                {alltags.length > 1 && <Select allOptions={alltags} label="Tags" />}
            </div>
        </div>
    );
};

export default PhotoCard;
