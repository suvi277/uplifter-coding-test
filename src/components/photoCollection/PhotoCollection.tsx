import React, { useEffect } from 'react';
import Loader from '../loader/Loader';
import PhotoCard from '../photoCard/PhotoCard';
import './PhotoCollection.scss';
import { Photo } from '../../types/Photo';

interface Props {
    status: string;
    photos?: Array<Photo>;
    getPhotos: any;
}

const PhotoCollectionComponent: React.FC<Props> = ({ photos = [], status, getPhotos }) => {
    useEffect(() => {
        getPhotos();
    }, [getPhotos]);

    return (
        <div className="container pt-5">
            {status === 'loading' ? (
                <div className="loader-container">
                    <Loader />
                </div>
            ) : (
                status === 'loaded' && (
                    <div className="d-flex flex-wrap">
                        {photos.map((item, index) => (
                            <PhotoCard key={index} {...item} />
                        ))}
                    </div>
                )
            )}
            {status === 'error' && <div className="error-content">Oops there is an error! Try Again later</div>}
        </div>
    );
};

export default PhotoCollectionComponent;
