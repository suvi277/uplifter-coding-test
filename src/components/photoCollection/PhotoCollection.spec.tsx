import React from 'react';
import { shallow, mount } from 'enzyme';
import PhotoCollection from './PhotoCollection';
import Loader from '../loader/Loader';
import PhotoCard from '../photoCard/PhotoCard';

describe('<PhotoCollection /> with default props', () => {
    const defaultProps = {
        status: 'loading',
        getPhotos: jest.fn(),
    };
    const wrapper = shallow(<PhotoCollection {...defaultProps} />);
    it('renders as expected', () => {
        expect(wrapper).toMatchSnapshot();
    });
    it('should render Loader as status is loading', () => {
        expect(wrapper.exists(Loader)).toBeTruthy();
    });

    describe('useEfect', () => {
        const Comp = () => {
            PhotoCollection({ ...defaultProps });
            return <div />;
        };

        const wrapper = mount(<Comp />);

        it('should call the `getPhotos` when mounted ', () => {
            expect(defaultProps.getPhotos).toHaveBeenCalled();
        });
    });

    describe('<PhotoCollection /> with photos props', () => {
        const eachPhoto = {
            title: 'sample title',
            ownername: 'sample owner name',
            description: {
                _content: 'some content',
            },
            url_q: 'url/path',
            owner: 'sampleowner',
            id: 'some id',
            tags: '',
        };
        const defaultProps = {
            status: 'loaded',
            photos: [eachPhoto],
            getPhotos: jest.fn(),
        };
        const wrapper = shallow(<PhotoCollection {...defaultProps} />);

        it('should render the `PhotoCard` as status is loaded with photos ', () => {
            expect(wrapper.exists(PhotoCard)).toBeTruthy();
        });
        it('should render one `PhotoCard` as length of photos array is 1 ', () => {
            expect(wrapper.find('PhotoCard').length).toBe(1);
        });
    });

    describe('<PhotoCollection /> with error props', () => {
        const defaultProps = {
            status: 'error',
            getPhotos: jest.fn(),
        };
        const wrapper = shallow(<PhotoCollection {...defaultProps} />);
        it('should render Loader as status is loading', () => {
            expect(wrapper.find('.error-content').text()).toEqual('Oops there is an error! Try Again later');
        });
    });
});
