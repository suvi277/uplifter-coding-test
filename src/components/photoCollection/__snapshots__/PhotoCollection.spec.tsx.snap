// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhotoCollection /> with default props renders as expected 1`] = `ShallowWrapper {}`;

exports[`<PhotoCollection /> with error props renders as expected 1`] = `ShallowWrapper {}`;

exports[`Photo Collection Component renders as expected 1`] = `ShallowWrapper {}`;

exports[`StreamCard Component renders as expected 1`] = `ShallowWrapper {}`;
